File: src/prod/src/data/tstore/MetadataManagerFileProperties.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"
#define METADATAMANAGERFP_TAG 'pfDM'

using namespace Data::TStore;
using namespace Data::Utilities;

MetadataManagerFileProperties::MetadataManagerFileProperties()
   : fileCount_(0),
   checkpointLSN_(-1),
   metadataHandleSPtr_(nullptr)
{
}

MetadataManagerFileProperties::~MetadataManagerFileProperties()
{
}

bool MetadataManagerFileProperties::Test_IgnoreCheckpointLSN = false;

NTSTATUS MetadataManagerFileProperties::Create(
   __in KAllocator& allocator,
   __out MetadataManagerFileProperties::SPtr& result)
{
   result = _new(METADATAMANAGERFP_TAG, allocator) MetadataManagerFileProperties();

   if (result == nullptr)
   {
      return STATUS_INSUFFICIENT_RESOURCES;
   }

   if (!NT_SUCCESS(result->Status()))
   {
      // Null Result while fetching failure status with no extra AddRefs or Releases
      return (SPtr(Ktl::Move(result)))->Status();
   }

   return STATUS_SUCCESS;
}

void MetadataManagerFileProperties::Write(__in BinaryWriter& writer)
{
   KInvariant(metadataHandleSPtr_ != nullptr);
   writer.Write(static_cast<ULONG32>(PropertyId::MetadataHandle));
   VarInt::Write(writer, static_cast<ULONG32>(BlockHandle::SerializedSize()));
   metadataHandleSPtr_->Write(writer);

   // FileCount
   writer.Write(static_cast<ULONG32>(PropertyId::DataFileCount));
   VarInt::Write(writer, static_cast<ULONG32>(sizeof(ULONG32)));
   writer.Write(static_cast<ULONG32>(fileCount_));

   if (Test_IgnoreCheckpointLSN)
   {
       return;
   }

   // 'CheckpointLSN' - LONG64
   writer.Write(static_cast<ULONG32>(PropertyId::CheckpointLSN));
   VarInt::Write(writer, static_cast<ULONG32>(sizeof(LONG64)));
   writer.Write(checkpointLSN_);
}

void MetadataManagerFileProperties::ReadProperty(
   __in BinaryReader& reader,
   __in ULONG32 property,
   __in ULONG32 valueSize)
{
   switch (static_cast<PropertyId::Enum>(property))
   {
   case PropertyId::MetadataHandle:
   {
      BlockHandle::SPtr metadataHandleSPtr = BlockHandle::Read(reader, GetThisAllocator());
      MetadataHandleSPtr = *metadataHandleSPtr;
      break;
   }

   case PropertyId::DataFileCount:
   {
      ULONG32 fileCount = 0;
      reader.Read(fileCount);
      FileCount = fileCount;
      break;
   }

   case PropertyId::CheckpointLSN:
   {
      LONG64 checkpointLSN = 0;
      reader.Read(checkpointLSN);
      CheckpointLSN = checkpointLSN;
      break;
   }

   default:
      FilePropertySection::ReadProperty(reader, property, valueSize);
      break;
   }
}

File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/ReplicatedStore/Disk/MetadataManagerFileProperties.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Store
{
    using System.IO;
    using Microsoft.ServiceFabric.Replicator;

    /// <summary>
    /// Properties for the MetadataManager.
    /// </summary>
    internal sealed class MetadataManagerFileProperties : FilePropertySection
    {
        internal static bool Test_IgnoreCheckpointLSN = false;

        /// <summary>
        /// If it does not exist, it is set to invalid.
        /// </summary>
        private long checkpointLSN = DifferentialStoreConstants.InvalidLSN;

        private enum PropertyId : int
        {
            MetadataHandle = 1,
            FileCount = 2,
            CheckpointLSN = 3,
        }

        public BlockHandle MetadataHandle { get; set; }

        public int FileCount { get; set; }

        public long CheckpointLSN
        {
            get
            {
                return this.checkpointLSN;
            }
            set
            {
                this.checkpointLSN = value;
            }
        }

        /// <summary>
        /// Serialize the <see cref="MetadataManagerFileProperties"/> into the given stream.
        /// </summary>
        /// <param name="writer">The stream to write to.</param>
        public override void Write(BinaryWriter writer)
        {
            // 'MetadataHandle' - BlockHandle
            writer.Write((int) PropertyId.MetadataHandle);
            VarInt.Write(writer, (int) BlockHandle.SerializedSize);
            this.MetadataHandle.Write(writer);

            // 'FileCount' - int
            writer.Write((int) PropertyId.FileCount);
            VarInt.Write(writer, (int) sizeof(int));
            writer.Write((int) this.FileCount);

            if (Test_IgnoreCheckpointLSN)
            {
                return;
            }

            // 'CheckpointLSN' - long
            writer.Write((int)PropertyId.CheckpointLSN);
            VarInt.Write(writer, (int)sizeof(long));
            writer.Write(this.checkpointLSN);
        }

        /// <summary>
        /// Read the given property value.
        /// </summary>
        /// <param name="reader">Stream to read from.</param>
        /// <param name="property">Property to read.</param>
        /// <param name="valueSize">The size in bytes of the value.</param>
        protected override void ReadProperty(BinaryReader reader, int property, int valueSize)
        {
            switch ((PropertyId) property)
            {
                case PropertyId.MetadataHandle:
                    this.MetadataHandle = BlockHandle.Read(reader);
                    break;

                case PropertyId.FileCount:
                    this.FileCount = reader.ReadInt32();
                    break;

                case PropertyId.CheckpointLSN:
                    this.checkpointLSN = reader.ReadInt64();
                    break;

                default:
                    base.ReadProperty(reader, property, valueSize);
                    break;
            }
        }
    }
}